db/db_impl/db_impl_write.cc

```diff
@@ -830,6 +830,17 @@ Status DBImpl::WriteImplWALOnly(
   return status;
 }
 
+void DBImpl::WriteStatusCheckOnLocked(const Status& status) {
+  // Is setting bg_error_ enough here?  This will at least stop
+  // compaction and fail any further writes.
+  // Caller must hold mutex_.
+  mutex_.AssertHeld();
+  if (immutable_db_options_.paranoid_checks && !status.ok() &&
+      !status.IsBusy() && !status.IsIncomplete()) {
+    error_handler_.SetBGError(status, BackgroundErrorReason::kWriteCallback);
+  }
+}
+
 void DBImpl::WriteStatusCheck(const Status& status) {
   // Is setting bg_error_ enough here?  This will at least stop
   // compaction and fail any further writes.
```

Comment:
Minor - can we acquire mutex here and call ```WriteStatusCheckOnLocked``` so the code is not duplicated?

Reply:
@anand1976 Updated and also addressing the deadlock in UnorderedWriteMemtable.

Reply:
I worry about this fix maybe cause performance problems. I mean, we must to acquire the mutex even if the if condition is not met. the condition is ```immutable_db_options_.paranoid_checks && !status.ok() && !status.IsBusy() && !status.IsIncomplete()```

Reply:
That's a good point. It can cause performance regression for unordered writes. I'm ok with reverting this function back to the way it was.